### DISCUSSION TITLE
"is implementing the internal SPI protocol-mapper"

### DISCUSSION THREAD

**Discussion by christlang** [keycloak/keycloak#9974] [Q&A]
### Description

I want to implement a Protocol Mapper to extend claims from keycloak.

Example code you can find under: https://github.com/christlang/spring_oauth2_examples/pull/6

I have to provide a jboss-deployment-structure.xml and define a module org.keycloak.keycloak-services
(https://github.com/christlang/spring_oauth2_examples/pull/6/files#diff-0eef7429069d66f7d1d325d9019ade23e9aa25d3b8d8b69919691bd07f6a3ef1)

When I deploy my protocol-mapper I get the following error in log
```
keycloak        | 10:02:53,725 WARN  [org.keycloak.services] (MSC service thread 1-5) KC-SERVICES0047: oidc-service-c-protocol-mapper (de.synyx.cl.oauth.example.keycloak.ExternalAttributeMapper) is implementing the internal SPI protocol-mapper. This SPI is internal and may change without notice
```

This kind of messages makes me nervous. Is there a better module to use?

### Motivation

I want to know the right way to implement a protocol mapper

**Comment by arnault01** [comment]
The warning message you are seeing can make nervous but can safely be ignored. It just states that the private SPI implemented can be stopped to be supported without notice.
The good thing with this warning message is that it tells you that it detected your protocol-mapper, so it is correctly configured.
